services/src/main/java/io/grpc/services/CallMetricRecorder.java

```diff
@@ -0,0 +1,102 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.services;
+
+import com.google.common.collect.ImmutableMap;
+import io.grpc.Context;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Utility to record call metrics for load-balancing. One instance per call.
+ */
+@ThreadSafe
+public final class CallMetricRecorder {
+  private static final CallMetricRecorder NOOP = new CallMetricRecorder().disable();
+
+  static final Context.Key<CallMetricRecorder> CONTEXT_KEY =
+      Context.key("io.grpc.services.CallMetricRecorder");
+
+  private final AtomicReference<ConcurrentHashMap<String, Double>> metrics =
+      new AtomicReference<>();
+  private volatile boolean disabled;
+
+  CallMetricRecorder() {
+  }
+
+  /**
+   * Returns the call metric recorder attached to the current context.  If there is none, returns
+   * a no-op recorder.
+   *
+   * <p><strong>IMPORTANT:</strong>It returns the recorder specifically for the current RPC call.
+   * <b>DO NOT</b> save the returned object or share it between different RPC calls.
+   *
+   * <p><strong>IMPORTANT:</strong>It must be called under the {@link Context} under which the RPC
+   * handler was called.  If it is called from a different thread, the Context must be propagated to
+   * the same thread, e.g., with {@link Context#wrap(Runnable)}.
+   */
+  public static CallMetricRecorder getCurrent() {
+    CallMetricRecorder recorder = CONTEXT_KEY.get();
+    return recorder != null ? recorder : NOOP;
+  }
+
+  /**
+   * Records a call metric measurement.  If RPC has already finished, this method is no-op.
+   *
+   * <p>A latter record will overwrite its former name-sakes.
+   *
+   * @return this recorder object
+   */
+  public CallMetricRecorder recordCallMetric(String name, double value) {
+    if (disabled) {
+      return this;
+    }
+    if (metrics.get() == null) {
+      // The chance of race of creation of the map should be very small, so it should be fine
+      // to create these maps that might be discarded.
+      metrics.compareAndSet(null, new ConcurrentHashMap<String, Double>());
+    }
+    metrics.get().put(name, value);
+    return this;
+  }
+
+  /**
+   * Returns all save metric values. No more metric values will be recorded after this method is
+   * called. Calling this method multiple times returns the same collection of metric values.
+   *
+   * @return a map containing all saved metric name-value pairs.
+   */
+  public Map<String, Double> snapshot() {
```

Comment:
Application code is not supposed to call this. Please make it package-private and add a public `CallMetricAccessor` that can access this method. Internally we can control the visibility of `CallMetricAccessor` to make sure applications can't call it.

`snapshot()` usually implies you can do multiple times. Here it is more like a `finalizeAndDump()`.

Reply:
From the API perspective, should we allow it to be called for multiple time?

Reply:
We should certainly allow it to be called multiple times. I should've said that `snapshot()` usually means you can mutate the data afterwards and get a different snapshot if called again, which is not the case here.
Because the data "freezes" in this method, a copy is not necessary. You can return a `Collections.unmodifiableMap(savedMetrics)`.

Reply:
Done.